src/qibo/derivative.py

```diff
@@ -0,0 +1,125 @@
+        initial_state ((1, 2**nqubits) matrix): initial state on which we act with the circuit.
+
+    Returns:
+        np.float value of the derivative of the expected value of H on the final state obtained applying U
+        to the initial state with respect to the target variational parameter.
+
+    Example:
+        .. testcode::
+
+            import qibo
+            import numpy as np
+            from qibo import hamiltonians, gates
+            from qibo.models import Circuit
+
+            # in order to see the difference with tf gradients
+            import tensorflow as tf
+            qibo.set_backend('tensorflow')
+
+            # defining an observable
+            def hamiltonian(nqubits = 1):
+                m0 = (1/nqubits)*hamiltonians.Z(nqubits).matrix
+                ham = hamiltonians.Hamiltonian(nqubits, m0)
+                return ham
+
+            # defining a dummy circuit
+            def circuit(nqubits = 1):
+                c = Circuit(nqubits = 1)
+                c.add(gates.RY(q = 0, theta = 0))
+                c.add(gates.RX(q = 0, theta = 0))
+                c.add(gates.M(0))
+                return c
+
+            # using GradientTape to benchmark
+            def gradient_tape(params):
+                params = tf.Variable(params)
+
+                with tf.GradientTape() as tape:
+                    c = circuit(nqubits = 1)
+                    c.set_parameters(params)
+                    h = hamiltonian()
+                    expected_value = h.expectation(c.execute().state())
+
+                grads = tape.gradient(expected_value, [params])
+                return grads
+
+            # initializing the circuit
+            c = circuit(nqubits = 1)
+
+            # some parameters
+            test_params = np.random.randn(2)
+            c.set_parameters(test_params)
+
+            # a test hamiltonian
+            test_hamiltonian = hamiltonian()
+
+            # running the psr with respect to the two parameters
+            grad_0 = parameter_shift(circuit = c, observable = test_hamiltonian, parameter_index = 0, gate_eigenv = 0.5)
+            grad_1 = parameter_shift(circuit = c, observable = test_hamiltonian, parameter_index = 1, gate_eigenv = 0.5)
+
+            # evaluating the gradients using tensorflow
+            tf_grads = gradient_tape(test_params)
+
+            print('Test gradient with respect params[0] with PSR: ', grad_0.numpy())
+            print('Test gradient with respect params[0] with tf:  ', tf_grads[0][0].numpy())
+            print('Test gradient with respect params[1] with PSR: ', grad_1.numpy())
+            print('Test gradient with respect params[1] with tf:  ', tf_grads[0][1].numpy())
+
+    """
+
+    if parameter_index > len(circuit.get_parameters()):
+        raise ValueError(
+            """This index doesn't exist! Use one of the circuit's parameters."""
+        )
```

Comment:
In qibo we have a custom way of raising errors in order to be captured by our logger. This is used as 
```suggestion
        from qibo.config import raise_error
        raise_error(ValueError, "This index doesn't exist! Use one of the circuit's parameters.")
```

Reply:
Out of curiosity: what is this logger used to?

Reply:
In practice it provides a uniform format for errors and standard output, which looks like `[Qibo 0.1.9.dev0|ERROR|2022-10-24 12:54:00]`. This may help when reporting issues as you can see the date and version. There may be some additional usage that I am not aware of.

The implementation of this format is done [here](https://github.com/qiboteam/qibo/blob/8b0e0d49cbab6eae7d550ae55081af9c0c0f448e/src/qibo/config.py#L82).

Reply:
I see, thanks!